### ISSUE TITLE
Improving the request methods to make synchronous by default

### ISSUE THREAD

**Issue by pickypg** [elastic/elasticsearch-groovy#21]
This stems from a likely change that will arrive in Elasticsearch 2.0 from https://github.com/elasticsearch/elasticsearch/issues/9201, where asynchronous client methods will be differentiated from synchronous variants by adding `Async` to the end of the asynchronous overload (thus making it not an overload anymore).

With respect to the Groovy client, all requests would be impacted by this decision, but it _could_ be hidden away to hide it in order to leave the Groovy client's API unchanged. I do _not_ want to leave it unchanged and this change will make it in sooner _or later_.

I would like to discuss potentially bringing this change into the Groovy client earlier by changing the existing methods from the current format:

```
ListenableActionFuture<Response> response = client.request { /* ... */ }
```

to

```
ListenableActionFuture<Response> response = client.requestAsync { /* ... */ }
```

and adding/changing a synchronous by default option as (notice only the return type differs from the current way):

```
Response response = client.request { /* ... */ }
```

If people are okay with this change, then I would be happy to make the change in the 1.5 release of the Groovy client. If people are really happy with the old asynchronous-by-default nature of the API, or afraid of the backwards incompatible change, then I can wait to push this into 2.0. I expect that most people currently use the API synchronously by immediately calling `actionGet` and I would like input on usage to see if I am right or wrong.

```
// expected most common usage currently:
Response response = client.request { /* ... */ }.actionGet()
```

**Comment by pickypg** [comment]
This will wait until `elasticsearch-groovy-2.0.0`.

**Comment by pickypg** [comment]
My current plans are for pre-2.0:

1. Change nothing in 1.4.
2. Create non-deprecated variants that take the form of `requestSomethingAsync` from existing methods in the form of `requestSomething` starting in 1.5+.
3. Deprecate all existing client methods in 1.5+.

My current plans for 2.0:

1. Replace deprecated methods with synchronous versions.
2. Maintain `*Async` variants.

**Comment by pickypg** [comment]
This change will not make it into the Groovy client beginning with ES 1.5 to ensure that I am not jumping the gun with the API change.

To be clear, as upon reflection I feel that it is not as clear as it could have been, the current API offers only asynchronous `Client` methods that look like this (this matches the Java client, except that the Java client only returns `ActionFuture<Response>`):

```groovy
ListenableActionFuture<Response> request(Request request);
```

However, the assumption is that most users want simple, synchronous methods. In order to add that behavior without goofy overloads, the above needs to be replaced. At the end, I _expect_ the API to look like:

```groovy
Response request(Request request);
ListenableActionFuture<Response> requestAsync(Request request);
```

However, in favor of not diverging from the Java API (it could go in a different direction), I am going to hold off on pushing this into a release branch.

**Comment by pickypg** [comment]
As we rapidly approach the first beta of Elasticsearch 2.0.0, I have changed the release to have a complementary, synchronous method that does _not_ change the current approach at all.

```groovy
Response requestSync(Request request);                          // synchronous
ListenableActionFuture<Response> request(Request request);      // asynchronous
ListenableActionFuture<Response> requestAsync(Request request); // asynchronous
```

The long term goal is still to replace the default option with a synchronous one and my hope is that the explicit methods will enable a simple transition off, then back onto the changed default.

Closed by 1e1eef61f58d832a6a01a5b2d9442580ee27d7b9
